Fix subset dropdown

jerboa88 opened the pull request:
Fixes for the subset dropdown as it doesn't work currently:
- Changed event handler as it was calling the wrong one and doing nothing
- Updated event handler to work with search functionality
- Created a local storage key for the subset value as previously it was `undefined`

I also added an `all` entry to the dropdown so that users can view fonts from all subsets, but I left that change out of this PR because I'm not sure if that's something you want to include. If so, I can add a commit for it as it's only a couple of lines.

pamelafox reviewed (changes requested):
Adding 'all' sounds good, thanks.

Trying this out, I think it needs a call to loadVisibleFonts() at the bottom of the filterFonts() function so that the font names are displayed with their font family. 
The thing I'm forgetting is why the heck the code has multiple checks for "!= 'khmer'" as it seems to work okay when I remove those checks. I think you can remove those checks unless you can see why I put those in there.  (That way the Khmer fonts will also get displayed with their font-family set to the font).

jerboa88 commented:
Ah yes, I missed the call to `loadVisibleFonts()`. I removed the checks for the khmer subset and haven't seen any issues so I'm not sure what that was for either.

I added debouncing to the dropdown and searchbox because calling `loadVisibleFonts()` in the `filterFonts()` function made it 3-4x slower. It still ran okay on my machine but I'm sure that could have caused issues for ppl (I tried using `throttle()` for this but it calls the function immediately which isn't great when you're trying to type text).

pamelafox commented on popup.js:
```diff
@@ -85,6 +88,7 @@ function onFontsLoad(json, fromCache) {
     lscache.set(LS_FONTS_API, json, 60*12);
   }
   loadVisibleFonts();
```
I think loadVisibleFonts should get called after onsubsetchange/onsearchchange so that it doesn't unnecessary load fonts that are about to get hidden.

jerboa88 replied:
Something wasn't working correctly when I first tried to put them in the reverse order but you're right, it doesn't make sense to be calling `loadVisibleFonts()` so much. I'll look into it.

pamelafox commented on popup.js:
```diff
@@ -85,6 +88,7 @@ function onFontsLoad(json, fromCache) {
     lscache.set(LS_FONTS_API, json, 60*12);
   }
   loadVisibleFonts();
+  onSubsetChange();
```
One consequence of the current code flow is that for the initial load, it ends up re-storing the two localstorage values unnecessarily and calling loadVisibleFonts three times. An alternate approach is to pass an isInitial boolean into those functions so as to avoid doing unneeded work.

jerboa88 replied:
Sure, good idea.

jerboa88 replied:
I just used the event passed to these functions via the event listeners to determine whether it's the initial call or not (rather than creating an extra flag).

Also replaced `loadVisibleFonts()` with `filterFonts()` because we might as well avoid doing the filtering twice as well.

pamelafox replied:
Ah good idea on checking the event!

pamelafox commented on popup.js:
```diff
@@ -126,6 +128,21 @@ function addRowBehavior(div) {
   };
 }
 
+// Returns a debounced version of the provided function
```
Nit, can you use /* style for this comment like the other function level comments?

pamelafox commented on popup.js:
```diff
@@ -39,8 +41,8 @@ function onLoad() {
-  document.getElementById("fonts-subset").addEventListener("change", onOptionChange);
-  document.getElementById("fonts-searchbox").addEventListener("input", onSearchChange);
+  document.getElementById("fonts-subset").addEventListener("change", (() => debounce(onSubsetChange))());
+  document.getElementById("fonts-searchbox").addEventListener("input", (() => debounce(onSearchChange))());
```
I don't think there's a need to use an IIFE here, right?
(IIFE: https://developer.mozilla.org/en-US/docs/Glossary/IIFE)
I think it can just be:
  document.getElementById("fonts-subset").addEventListener("change", debounce(onSubsetChange));
  document.getElementById("fonts-searchbox").addEventListener("input", debounce(onSearchChange));

pamelafox reviewed (changes requested):
Thanks for the changes, just one more thing
